### DISCUSSION TITLE
Stufenschalter mit Messung zeigt dauerhaft Verbrauch im Sunnyportal

### DISCUSSION THREAD

**Discussion by patrick-tix** [camueller/SmartApplianceEnabler#366] [Allgemein]
Hallo,

ich habe gestern einen Stufenschalter für meinen Heizstab angelegt.
Gemessen wird mit einem Shelly 3EM, geschaltet wird Stufe 1 über den Shelly3EM Ausgang, Stufe 2 und 3 über je einen Shelly 1.

Unter Status im SAE 2.0.6 wird der Heizstab nun mit dauerhaft mit blinkender grüner LED dargestellt, obwohl die Shellys ausgeschaltet sind. Im Portal wird zudem in der Verbraucherbilanz eine dauerhafte Leistung von 1500W angezeigt.

Was hab ich hier verkehrt konfiguriert?
![heizstab_sae_status](https://user-images.githubusercontent.com/115613948/196118930-9478c401-7418-43fe-b61a-87a42080c187.png)

![heizstab_sae_geraet](https://user-images.githubusercontent.com/115613948/196118929-18542903-9502-45db-a68d-3b2e64a9254c.png)

![heizstab_sae_zaehler](https://user-images.githubusercontent.com/115613948/196118924-d4e7b6c0-d1b6-4752-ac8c-100cc3bff4c9.png)

![heizstab_sae_schalter_1](https://user-images.githubusercontent.com/115613948/196118922-12600ce8-a0e2-492e-8485-214831eeb887.png)

![heizstab_sae_schalter_2](https://user-images.githubusercontent.com/115613948/196118921-62fbc9fb-7940-4724-b0bb-1d1de2fdbdfc.png)

![heizstab_sae_schalter_3](https://user-images.githubusercontent.com/115613948/196118920-a6d0a06c-81ae-4e37-b7fa-3c9a3ab020af.png)

![heizstab_sae_schalter_leistungsstufen](https://user-images.githubusercontent.com/115613948/196118917-91fd746a-9089-4d97-b8e8-9bec8a67bfa6.png)

![heizstab_sae_schaltzeiten](https://user-images.githubusercontent.com/115613948/196118914-deffab8b-988b-4ad7-bebc-5b2f09dd6a92.png)

**Comment by camueller** [comment]
Möglicherweise ist ein Implementierungsfehler die Ursache. Kannst Du mir ein Log-File bereitstellen (Google Drive, ...)?

**Comment by patrick-tix** [comment]
Hallo,

ich habe einmal das LOG von heute sowie die aktuelle semp.xml bei Google Drive hochgeladen:

[https://drive.google.com/drive/folders/1A2LRjspfIFav3B2vEK2EoaV-uW1_CHD0?usp=sharing](url)

**Comment by Lausbua** [comment]
Eventuell liegt es an der fehlenden Angabe der min. Leistungsaufnahme in den Eigenschaften von deinem Gerät?

Laut Anleitung muss diese befüllt werden, https://github.com/camueller/SmartApplianceEnabler/blob/2.0/doc/LevelSwitch_DE.md

![image](https://user-images.githubusercontent.com/3309871/196282107-12f03092-1b65-47f2-b24c-8843ba17088d.png)

**Reply by camueller** [reply]
Das stimmt, bei Verbrauchern mit variabler Leistungsaufnahme muss die min. Leistungsaufnahme angegeben werden. Das wird leider aktuell im Frontend nicht korrekt validiert.

**Comment by patrick-tix** [comment]
Ich habe die minimale Leistungsaufnahme heute morgen eingetragen (entsprechend der Stufe 1). Bisher werden dem Portal aber immernoch 1500W Leistungsaufnahme übermittelt. Ich starte den SAE mal neu und berichte weiter.

**Comment by patrick-tix** [comment]
Nach dem Neustart wird zumindest nicht mehr der dauerhafte Verbrauch im Portal angezeigt.

Jedoch wird der Heizstab (noch) nicht eingeplant, obwohl genug Überschuss vorhanden ist. Außerdem blinkt die Ampel im SAE immer noch dauerhaft.

Kann ich noch etwas prüfen?

Die aktuellen Logs habe ich nochmal hochgeladen:
https://drive.google.com/drive/folders/1A2LRjspfIFav3B2vEK2EoaV-uW1_CHD0?usp=sharing

Vielen Dank vorab.

**Reply by patrick-tix** [reply]
Einen Fehler habe ich gerade schon gefunden:
Ich habe beim Zählerstand den falschen JSON Path angegeben (vom Tasmota statt vom Shelly).
Kann das die Ursache sein?

**Reply by patrick-tix** [reply]
Ich habe noch den HTTP Status für die Schalter entfernt. Jetzt blinkt die Ampel gelb.

Ich parse den HTTP-Status derzeit mit Regex. Geht das auch mit JSON? Wenn ja, wie?

Vielen Dank.

**Comment by camueller** [comment]
> Einen Fehler habe ich gerade schon gefunden: Ich habe beim Zählerstand den falschen JSON Path angegeben (vom Tasmota statt vom Shelly). Kann das die Ursache sein?

Wie soll sich der SAE korrekt verhalten, wenn der Zähler so konfiguriert ist, dass er bei jeder Abfrage einen Fehler hat?
Im Log-File steht:
```
2022-10-18 12:19:43,152 ERROR [Timer-0] d.a.s.u.GuardedTimerTask [GuardedTimerTask.java:60] F-20181211-000000000006-00: Error executing timer task name=PollEnergyMeter id=32944369
com.jayway.jsonpath.PathNotFoundException: Missing property in path $['StatusSNS']
```

Wenn man die Response von Deinem Shelly mal formatiert, sieht sie so aus:
```
{
  "wifi_sta": {
    "connected": true,
    "ssid": "WLAN Keller",
    "ip": "192.168.2.35",
    "rssi": -44
  },
  "cloud": {
    "enabled": true,
    "connected": true
  },
  "mqtt": {
    "connected": false
  },
  "time": "12:19",
  "unixtime": 1666088383,
  "serial": 47,
  "has_update": false,
  "mac": "34945474A88F",
  "cfg_changed_cnt": 0,
  "actions_stats": {
    "skipped": 0
  },
  "relays": [
    {
      "ison": false,
      "has_timer": false,
      "timer_started": 0,
      "timer_duration": 0,
      "timer_remaining": 0,
      "overpower": false,
      "is_valid": true,
      "source": "cloud"
    }
  ],
  "emeters": [
    {
      "power": 0,
      "pf": -0.02,
      "current": 0.01,
      "voltage": 229.74,
      "is_valid": true,
      "total": 2164.7,
      "total_returned": 2.3
    },
    {
      "power": 0,
      "pf": -0.05,
      "current": 0.01,
      "voltage": 229.12,
      "is_valid": true,
      "total": 2107.8,
      "total_returned": 2.2
    },
    {
      "power": 0,
      "pf": -0.01,
      "current": 0.01,
      "voltage": 230.88,
      "is_valid": true,
      "total": 2165.3,
      "total_returned": 2.3
    }
  ],
  "total_power": 0,
  "emeter_n": {
    "current": 0,
    "ixsum": 0.02,
    "mismatch": false,
    "is_valid": false
  },
  "fs_mounted": true,
  "v_data": 1,
  "ct_calst": 0,
  "update": {
    "status": "idle",
    "has_update": false,
    "new_version": "20220830-080542/v1.12-3EM-gcf4f7c2",
    "old_version": "20220830-080542/v1.12-3EM-gcf4f7c2"
  },
  "ram_total": 49928,
  "ram_free": 31876,
  "fs_size": 233681,
  "fs_free": 154867,
  "uptime": 148045
}
```
Du solltest also den Zähler wie folgt konfigurieren:
-  Format "JSON"
-  Parameter "Leistung"
- "Pfad für Extraktion" etwa `$.emeters[1].power`, wobei Du die Zahl durch den Index des jeweiligen Zählers [0, 1, 2, ...] ersetzen musst

Due kannst den JSON-Path hier bestimmen lassen: https://jsonpathfinder.com/

Geld blinkende Ampel ist gut, d.h. es kann jederzeit eine Einschaltung für Überschuss-Energie erfolgen. Wobei die Ampel bei Stufenschaltern durchaus noch Fehler haben kann. Massgeblich ist, was an der SEMP-Schnittstelle gemeldet wird (die Ampel ist "nur" für den User).

Einschaltungen solltest Du erst erwarten, wenn ansonsten alles fehlerfrei läuft - auch der Zähler.

**Comment by patrick-tix** [comment]
Danke für die Rückmeldung.

Der Heizstab war zu beginn 3phasig angeschlossen, aber da ich derzeit keine 4kW PV-Überschuss habe, wurde der Heizstab auf 3 einzelne Stufen umgebaut. Diese 3 Stufen wollte ich mit den Stufenschalter ansteuern und mit dem Shelly 3EM gesamt messen.

Leider gibt das Shelly 3EM keinen Gesamtzählerstand für alle 3 Phasen aus. Soll ich dann lieber Leistung und $.total_power nehmen, oder gibt es da noch eine andere Möglichkeit?

Ist es sinnvoll, mehrere Zeitpläne anzulegen, oder reicht ein Zeitplan von 8:00 Uhr bis 20:00 und der Heizstab schaltet dann mehrfach ein und aus, sobald der Überschuss passt und die max. Energie noch nicht erreicht ist?

**Reply by camueller** [reply]
> Leider gibt das Shelly 3EM keinen Gesamtzählerstand für alle 3 Phasen aus. Soll ich dann lieber Leistung und $.total_power nehmen, oder gibt es da noch eine andere Möglichkeit?

Ich kenne Deinen Shelly nicht, aber wenn ich mir das JSON anschaue, dann findet sich da `total_power` - das klingt passend. Das kannst Du ja selbst testen, ob dieses Feld den richtigen Wert beinhaltet. Du musst dann den JSON-Path anpassen wie oben beschrieben.

> Ist es sinnvoll, mehrere Zeitpläne anzulegen, oder reicht ein Zeitplan von 8:00 Uhr bis 20:00 und der Heizstab schaltet dann mehrfach ein und aus, sobald der Überschuss passt und die max. Energie noch nicht erreicht ist?

Ein Zeitplan sollte reichen.

**Comment by patrick-tix** [comment]
Ich sehe gerade im Portal unter Prognose, dass gegen Mittag ein Verbrach von 23 kWh erwartet wird.

Unter Schaltzeiten habe ich eine Max. Energie von 20.000 Wh angegeben. Muss ich hier evtl. einen anderen Wert angeben, oder ist dies nur eine wirkliche maximale Energiemenge und hat nichts mit dem Ein- und Ausschaltverhalten zu tun?

**Reply by camueller** [reply]
> Ich sehe gerade im Portal unter Prognose, dass gegen Mittag ein Verbrach von 23 kWh erwartet wird.

Ignoriere bitte das Portal erstmal - das ist erst interessant, wenn im SAE alles richtig läuft!!

Entscheidend ist, was der SAE unter /semp an den SHM übermittelt (Leistung, Zeitpläne, ...)


> Unter Schaltzeiten habe ich eine Max. Energie von 20.000 Wh angegeben. Muss ich hier evtl. einen anderen Wert angeben, oder ist dies nur eine wirkliche maximale Energiemenge und hat nichts mit dem Ein- und Ausschaltverhalten zu tun?

Wenn Du Überschussenergie willst, musst Du auch ein min. Energie (meist "0") angeben (siehe [Doku](https://github.com/camueller/SmartApplianceEnabler/blob/master/doc/Schedules_DE.md) - wobei das in Deinem Fall für `Energiemenge` gilt, nicht für `Laufzeit`).

**Comment by patrick-tix** [comment]
Ich habe den Zähler einmal angepasst und  die Energiemenge erstmal zwischen 0 und 2000 eingestellt.
Wie kann ich nun prüfen, ob vom SAE erstmal alles richtig läuft?

Aktuelle LOG und SEMP Ausgabe habe ich auch nochmal hochgeladen:

https://drive.google.com/drive/folders/1A2LRjspfIFav3B2vEK2EoaV-uW1_CHD0?usp=sharing

Vielen Dank nochmal für die Unterstützung.

**Comment by camueller** [comment]
Im Log-File sind keine Fehler - das ist schon mal gut!

Im `/semp` findet sich:
```
       <Characteristics>
            <MaxPowerConsumption>1500</MaxPowerConsumption>
            <MinPowerConsumption>500</MinPowerConsumption>
            <MinOnTime>300</MinOnTime>
            <MinOffTime>300</MinOffTime>
            <PowerLevels>
                <PowerLevel>500</PowerLevel>
                <PowerLevel>1000</PowerLevel>
                <PowerLevel>1500</PowerLevel>
            </PowerLevels>
        </Characteristics>
```
... sieht gut aus. Ich würde aber die "Min. Einschaltzeit" und "Min. Ausschaltzeit" erstmal nicht setzen, da meiner Erfahrung nach dann der SHM zögerlich mit dem Einschalten ist. Wenn alles läuft, kannst Du die Werte wieder setzen.

```
    <DeviceStatus>
        <DeviceId>F-20181211-000000000006-00</DeviceId>
        <EMSignalsAccepted>true</EMSignalsAccepted>
        <Status>Off</Status>
        <PowerConsumption>
            <PowerInfo>
                <AveragePower>0</AveragePower>
                <Timestamp>0</Timestamp>
                <AveragingInterval>60</AveragingInterval>
            </PowerInfo>
        </PowerConsumption>
    </DeviceStatus>
```
... passt auch und sagt, dass wir Schaltbefehle vom SHM akzeptieren und wir aktuell ausgeschaltet sind mit Leistungsaufnahme 0W.

```
       <Timeframe>
            <DeviceId>F-20181211-000000000006-00</DeviceId>
            <EarliestStart>0</EarliestStart>
            <LatestEnd>22420</LatestEnd>
            <MinEnergy>0</MinEnergy>
            <MaxEnergy>2000</MaxEnergy>
        </Timeframe>
```
... passt auch und bedeutet, dass der SHM sofort einen Einschaltbefehl senden, da wir 2000 Wh Überschussenergie aufnehmen können.

Alles in allem sieht das gut aus und sollte dazu führen, dass Du bei genügen Überschuss einen Einschaltbefehl vom SHM bekommst.

Wie gross ist eigentlich Deine PV-Anlage? Gibt es eine Batterie, die auch geladen werden muss?

**Comment by patrick-tix** [comment]
Hallo,

ich habe heute morgen noch die Ein- und Ausschaltzeit rausgenommen und nun hat es geklappt. Stufe 1 und 2 sind gerade eingeschaltet worden.

Vielen vielen Dank

**Reply by camueller** [reply]
> ich habe heute morgen noch die Ein- und Ausschaltzeit rausgenommen und nun hat es geklappt. Stufe 1 und 2 sind gerade eingeschaltet worden.

Sehr schön - so soll es sein :-)
